File: src/Features/VisualBasic/Completion/CompletionProviders/SymbolCompletionItemRules.vb
' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Imports Microsoft.CodeAnalysis.Completion
Imports Microsoft.CodeAnalysis.Completion.Providers

Namespace Microsoft.CodeAnalysis.VisualBasic.Completion.Providers
    Friend Class SymbolCompletionItemRules
        Inherits CompletionItemRules

        Public Shared ReadOnly Property Instance As New SymbolCompletionItemRules()

        Public Overrides Function IsCommitCharacter(completionItem As CompletionItem, ch As Char, textTypedSoFar As String) As Result(Of Boolean)
            Dim symbolItem = TryCast(completionItem, SymbolCompletionItem)
            If symbolItem IsNot Nothing AndAlso symbolItem.Context.IsInImportsDirective Then
                ' If the user is writing "Imports S" then the only commit character is <dot>
                ' as they might be typing an Imports alias.
                Return ch = "."c
            End If

            Return MyBase.IsCommitCharacter(completionItem, ch, textTypedSoFar)
        End Function

    End Class
End Namespace

File: src/Features/CSharp/Completion/CompletionProviders/SymbolCompletionItemRules.cs
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using Microsoft.CodeAnalysis.Completion;
using Microsoft.CodeAnalysis.Completion.Providers;

namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
{
    internal class SymbolCompletionItemRules : CompletionItemRules
    {
        public static SymbolCompletionItemRules Instance { get; } = new SymbolCompletionItemRules();

        public override Result<bool> IsCommitCharacter(CompletionItem completionItem, char ch, string textTypedSoFar)
        {
            var symbolItem = completionItem as SymbolCompletionItem;
            if (symbolItem != null && symbolItem.Context.IsInImportsDirective)
            {
                // If the user is writing "using S" then the only commit characters are <dot> and
                // <semicolon>, as they might be typing a using alias.
                return ch == '.' || ch == ';';
            }

            return base.IsCommitCharacter(completionItem, ch, textTypedSoFar);
        }
    }
}
